### DISCUSSION TITLE
Lost changes in MR due to Renovate rebase(?)

### DISCUSSION THREAD

**Discussion by p3k** [renovatebot/renovate#11332] [Request Help]
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

renovate@^25

### Describe the bug

We noticed something was not working as expected after resolving a Renovate MR. A quick check showed changes I applied to the corresponding branch were gone.

The assignee who merged the branch was not aware something is missing and thus, the changes missing went unnoticed.

This was the second time (we know of) that this happened. 

Because logging was not enabled I would need some advice how we could examine this issue further “post-mortem”.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste any log here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment]
Renovate can only detect changes to it's branches, if they are done with a different git author. If you reuse the same git author, renovate can't detect the manual changes and will do a force-push if the branch needs an update.

https://docs.renovatebot.com/self-hosted-configuration/#gitauthor

**Reply by p3k** [reply]
thanks for the hint, but we are using a special gitlab account for the renovate commits. thus, i assume this cannot explain our loss.

i also see a list with some mrs – which have additional changes in a similar fashion – in the “Edited/Blocked” section. obviously, generally renovate can and does detect changed branches. still, it *did* overwrite changes in some other mrs.

**Reply by rarkins** [reply]
Please run with LOG_LEVEL=debug for some time in case the problem reoccurs. Then you can find the log from when the commit happened and hopefully get some indication of logic prior to that. If you can gather more evidence like that then we can decide whether to file it as a bug issue again instead of a discussion.

**Reply by viceice** [reply]
i pretty sure someone checked the rebase checkbox or somebody did git ammend / force-push.
That way renovate will overwrite any changes.

Can't you see changes in the mr history?

Renovate also only checks the latest commit of branch, so if you reorder custom commit's before renovate commits, renovate can't detect the changes. So make sure, custom changes are always a seperate commit after renovate commit.

**Reply by viceice** [reply]
We've also official templates for gitlab which saves debug logs as job attachment

https://gitlab.com/renovate-bot/renovate-runner

**Reply by p3k** [reply]
> i pretty sure someone checked the rebase checkbox or somebody did git ammend / force-push.
> That way renovate will overwrite any changes.

although i cannot rule it out 100%, it is very unlikely. please do not make assumptions.

> Can't you see changes in the mr history?

i think i can see the changes in gitlab’s mr comparison – the red arrows point to renovate’s versions, currently visible are my changes that got lost in the latest version.

![image](https://user-images.githubusercontent.com/16631/130060431-c795fddc-034d-41a5-a5fd-d7e616979e18.png)

> Renovate also only checks the latest commit of branch, so if you reorder custom commit's before renovate commits, renovate can't detect the changes. So make sure, custom changes are always a seperate commit after renovate commit.

i doubt i did this in the branch, as you can see it does not contain a lot of commits, anyway.

**Reply by viceice** [reply]
OK, so the only way to find out more is to run with some debug log and check it regularly if this happens again 🤷‍♂️ 

I've not seen wrongly overwritten custom changes for a long time. 🤔

**Reply by p3k** [reply]
> OK, so the only way to find out more is to run with some debug log and check it regularly if this happens again man_shrugging

true, so i enabled debug logging now as suggested by you and @rarkins.

> I've not seen wrongly overwritten custom changes for a long time. thinking

well, that’s a good thing, isn’t it? nevertheless, something happened, and maybe it happens again and we can work it out. if not, i assume that’s ok, too ;)

thanks for the quick assistance @viceice and @rarkins

**Reply by rarkins** [reply]
The logic surrounding rebasing is complex so it's entirely possible there's still a bug buried somewhere. If you see it happening again and have the log, it would be great.
